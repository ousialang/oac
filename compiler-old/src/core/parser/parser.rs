/* This file is part of Oac.
 *
 * Oac is free software: you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later
 * version.
 *
 * Oac is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License along with
 * Oac.  If not, see <https://www.gnu.org/licenses/>. */

use core::parser::cst::Cst;

pub trait CharUtils {
    fn is_symbol(self) -> bool;
    fn is_opening_bracket(self) -> bool;
    fn is_closing_bracket(self) -> bool;
    fn is_right_side_of(self, c: char) -> bool;
}

impl CharUtils for char {
    fn is_symbol(self) -> bool {
        "/+*-!|%&=?^@#.:,<>".contains(self)
    }

    fn is_opening_bracket(self) -> bool {
        "{[(".contains(self)
    }

    fn is_closing_bracket(self) -> bool {
        "}])".contains(self)
    }

    fn is_right_side_of(self, c: char) -> bool {
        match (c, self) {
            ('(', ')') | ('[', ']') | ('{', '}') => true,
            _ => false,
        }
    }
}

impl Parser {
    fn new() -> Parser {
        Parser {
            cst: Some(Cst::new()),
            feedbacks: Vec::new(),
            pedigree_of_current_scope: String::new(),
            buffer: String::new(),
            flag: false,
        }
    }

    fn parse_word(&mut self, chars: &mut Peekable<Chars>) -> char {
        loop {
            let c = chars.next().unwrap();
            if c.is_alphabetic() {
                self.buffer.push(c);
            } else {
                return c;
            }
            if chars.peek().is_none() {
                return '\0';
            }
        }
    }

    fn parse_symbol(&mut self, chars: &mut Peekable<Chars>) -> char {
        loop {
            let c = chars.next().unwrap();
            if c.is_symbol() {
                self.buffer.push(c);
            } else {
                return c;
            }
            if chars.peek().is_none() {
                return '\0';
            }
        }
    }

    fn parse_number(&mut self, chars: &mut Peekable<Chars>) -> char {
        loop {
            let c = chars.next().unwrap();
            if c.is_digit(10) {
                self.buffer.push(c);
            } else {
                self.add_token(Token {
                    location: Range { start: 0, end: 1 },
                    kind: TokenKind::Number(Number {
                        buffer: "abc".to_owned(),
                    }),
                });
                return c;
            }
            if chars.peek().is_none() {
                return '\0';
            }
        }
    }

    fn parse_string(&mut self, chars: &mut Peekable<Chars>) -> char {
        loop {
            let c = chars.next().unwrap();
            if self.flag {
                self.buffer.push(c);
            } else if c == '"' {
                chars.next().unwrap_or('\0');
            } else if c == '/' {
                self.flag = true;
            } else {
                self.buffer.push(c);
            }
            if chars.peek().is_none() {
                self.flag = true;
                return '\0';
            }
        }
    }

    fn parse_opening_bracket(&mut self, chars: &mut Peekable<Chars>) {
        self.pedigree_of_current_scope.push(chars.next().unwrap());
    }

    fn parse_closing_bracket(&mut self, chars: &mut Peekable<Chars>) {
        match self.pedigree_of_current_scope.pop() {
            Some(c) => {
                if !chars.next().unwrap().is_right_side_of(c) {
                    // TODO
                }
            }
            None => (), // TODO
        }
    }

    fn parse_whitespace(&mut self, chars: &mut Peekable<Chars>) {}

    fn add_token(&mut self, t: Token) {}

    fn peek_position(&mut self, chars: &mut Peekable<Chars>) -> Position {
        let option_c = chars.peek();
        if option_c.is_none() {
            return Position::EOF;
        }
        let c = option_c.unwrap();
        let mut with_chars = WithChars { chars: chars };
        if c.is_whitespace() {
            return Position::Whitespace;
        } else if c.is_word() {
            return Position::Word;
        } else if c.is_opening_bracket() {
            return Position::LBracket;
        } else if c.is_closing_bracket() {
            return Position::RBracket;
        } else if c.is_symbol() {
            return Position::Symbol;
        } else if c.is_string() {
            return Position::String;
        } else if c.is_number() {
            return Position::Number;
        } else {
            return Position::Unknown;
        }
    }
}

pub trait CharUtils {
    fn is_symbol(self) -> bool;
    fn is_opening_bracket(self) -> bool;
    fn is_closing_bracket(self) -> bool;
    fn is_right_side_of(self, c: char) -> bool;
}

impl CharUtils for char {
    fn is_symbol(self) -> bool {
        "/+*-!|%&=?^@#.:,<>".contains(self)
    }

    fn is_opening_bracket(self) -> bool {
        "{[(".contains(self)
    }

    fn is_closing_bracket(self) -> bool {
        "}])".contains(self)
    }

    fn is_right_side_of(self, c: char) -> bool {
        match (c, self) {
            ('(', ')') | ('[', ']') | ('{', '}') => true,
            _ => false,
        }
    }
}

impl Parser {
    fn new() -> Parser {
        Parser {
            ast: Some(Ast::new()),
            feedbacks: Vec::new(),
            pedigree_of_current_scope: String::new(),
            buffer: String::new(),
            flag: false,
        }
    }

    fn parse_word(&mut self, chars: &mut Peekable<Chars>) -> char {
        loop {
            let c = chars.next().unwrap();
            if c.is_alphabetic() {
                self.buffer.push(c);
            } else {
                return c;
            }
            if chars.peek().is_none() {
                return '\0';
            }
        }
    }

    fn parse_symbol(&mut self, chars: &mut Peekable<Chars>) -> char {
        loop {
            let c = chars.next().unwrap();
            if c.is_symbol() {
                self.buffer.push(c);
            } else {
                return c;
            }
            if chars.peek().is_none() {
                return '\0';
            }
        }
    }

    fn parse_number(&mut self, chars: &mut Peekable<Chars>) -> char {
        loop {
            let c = chars.next().unwrap();
            if c.is_digit(10) {
                self.buffer.push(c);
            } else {
                self.add_token(Token {
                    location: Range { start: 0, end: 1 },
                    kind: TokenKind::Number(Number {
                        buffer: "abc".to_owned(),
                    }),
                });
                return c;
            }
            if chars.peek().is_none() {
                return '\0';
            }
        }
    }

    fn parse_string(&mut self, chars: &mut Peekable<Chars>) -> char {
        loop {
            let c = chars.next().unwrap();
            if self.flag {
                self.buffer.push(c);
            } else if c == '"' {
                chars.next().unwrap_or('\0');
            } else if c == '/' {
                self.flag = true;
            } else {
                self.buffer.push(c);
            }
            if chars.peek().is_none() {
                self.flag = true;
                return '\0';
            }
        }
    }

    fn parse_opening_bracket(&mut self, chars: &mut Peekable<Chars>) {
        self.pedigree_of_current_scope.push(chars.next().unwrap());
    }

    fn parse_closing_bracket(&mut self, chars: &mut Peekable<Chars>) {
        match self.pedigree_of_current_scope.pop() {
            Some(c) => {
                if !chars.next().unwrap().is_right_side_of(c) {
                    // TODO
                }
            }
            None => (), // TODO
        }
    }

    fn parse_whitespace(&mut self, chars: &mut Peekable<Chars>) {}

    fn add_token(&mut self, t: Token) {}

    fn peek_position(&mut self, chars: &mut Peekable<Chars>) -> Position {
        let option_c = chars.peek();
        if option_c.is_none() {
            return Position::EOF;
        }
        let c = option_c.unwrap();
        let mut with_chars = WithChars { chars: chars };
        if c.is_whitespace() {
            return Position::Whitespace;
        } else if c.is_word() {
            return Position::Word;
        } else if c.is_opening_bracket() {
            return Position::LBracket;
        } else if c.is_closing_bracket() {
            return Position::RBracket;
        } else if c.is_symbol() {
            return Position::Symbol;
        } else if c.is_string() {
            return Position::String;
        } else if c.is_number() {
            return Position::Number;
        } else {
            return Position::Unknown;
        }
    }
}

pub trait CharUtils {
    fn is_symbol(self) -> bool;
    fn is_opening_bracket(self) -> bool;
    fn is_closing_bracket(self) -> bool;
    fn is_right_side_of(self, c: char) -> bool;
}

impl CharUtils for char {
    fn is_symbol(self) -> bool {
        "/+*-!|%&=?^@#.:,<>".contains(self)
    }

    fn is_opening_bracket(self) -> bool {
        "{[(".contains(self)
    }

    fn is_closing_bracket(self) -> bool {
        "}])".contains(self)
    }

    fn is_right_side_of(self, c: char) -> bool {
        match (c, self) {
            ('(', ')') | ('[', ']') | ('{', '}') => true,
            _ => false,
        }
    }
}
