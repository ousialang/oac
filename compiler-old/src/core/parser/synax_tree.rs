/* This file is part of Oac.
 *
 * Oac is free software: you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later
 * version.
 *
 * Oac is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
 * details.
 *
 * You should have received a copy of the GNU General Public License along with
 * Oac.  If not, see <https://www.gnu.org/licenses/>. */

// cst.rs
// ======
//
// An incremental parser that builds the Concrete Syntax Tree (CST) of some source.

use std::collections::LinkedList;
use std::fs::File;
use std::io::prelude::*;
use std::iter::Peekable;
use std::ops::Range;
use std::path::PathBuf;
use std::str::Chars;

#[derive(Debug)]
pub struct Cst {
    source: String,
    tree: LinkedList<Token>,
    feedbacks: Vec<String>,
}

#[derive(Debug)]
pub struct Token {
    location: Range<u64>,
    kind: TokenKind,
}

#[derive(Debug)]
pub enum TokenKind {
    NestedScope(NestedScope),
    Number(Number),
    Literal(Literal),
    Word,
    Symbol,
}

#[derive(Debug)]
pub struct NestedScope {
    bracket: Bracket,
    contents: Vec<Token>,
}

#[derive(Debug)]
pub enum Bracket {
    Round,
    Square,
    Curly,
}

#[derive(Debug)]
pub struct Number {
    buffer: String,
}

#[derive(Debug)]
pub struct Literal {
    true_str: String,
}

impl Pass<SyntaxTree> for UncommentedLines {
    pub fn transform(self) -> Result<SyntaxTree, failure::Error>
        let mut parser = Parser::new();
        // TODO: create warnings when src is empty
        if src.is_empty() {
            return (parser.ast, parser.feedbacks);
        }
        let mut chars = src.chars().peekable();
        loop {
            match parser.peek_position(&mut chars) {
                Position::Whitespace(ref mut s) => parser.parse_whitespace(s.chars),
                Position::Symbol(ref mut s) => parser.parse_symbol(s.chars),
                Position::Word(ref mut s) => parser.parse_word(s.chars),
                Position::String(ref mut s) => parser.parse_string(s.chars),
                Position::Number(ref mut s) => parser.parse_number(s.chars),
                Position::LBracket(ref mut s) => parser.parse_opening_bracket(s.chars),
                Position::RBracket(ref mut s) => parser.parse_closing_bracket(s.chars),
                Position::EOF => {
                    parser.terminate();
                    break;
                }
                _ => println!("bvvzxcvxzc"),
            }
        }
        parser.ast
    }

    pub fn from_file(path: PathBuf) -> Option<Cst> {
        let mut buffer = String::new();
        File::open(path).unwrap().read_to_string(&mut buffer);
        Cst::from_string(buffer.as_str())
    }

    fn add_token(&mut self, t: Token) {
        self.tokens.push(t);
    }
}

enum Position<'a> {
    Whitespace(WithChars<'a>),
    String(WithChars<'a>),
    Number(WithChars<'a>),
    Word(WithChars<'a>),
    Symbol(WithChars<'a>),
    EOF,
    Unknown(WithChars<'a>),
    LBracket(WithChars<'a>),
    RBracket(WithChars<'a>),
}

struct WithChars<'a> {
    chars: &'a Peekable<Chars<'a>>,
}
