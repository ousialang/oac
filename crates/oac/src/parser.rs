use serde::Serialize;
use tracing::trace;

use crate::tokenizer::{TokenData, TokenList};

#[derive(Clone, Debug, Serialize)]
pub struct Ast {
    pub top_level_functions: Vec<Function>,
}

#[derive(Clone, Debug, Serialize)]
pub struct Function {
    pub name: String,
    pub parameters: Vec<Parameter>,
    pub body: Vec<Statement>,
}

#[derive(Clone, Debug, Serialize)]
pub enum Statement {
    Assign { variable: String, value: Expression },
    Return { expr: Expression },
    Expression { expr: Expression },
}

#[derive(Clone, Debug, Serialize)]
pub enum Expression {
    Literal(Literal),
    Variable(String),
    Call(String, Vec<Expression>),
}

#[derive(Clone, Debug, Serialize)]
pub enum Literal {
    Int(u32),
    String(String),
}

#[derive(Clone, Debug, Serialize)]
pub struct Parameter {
    pub name: String,
    pub ty: String,
}

fn parse_statement(tokens: &mut Vec<TokenData>) -> anyhow::Result<Statement> {
    trace!("Parsing statement: {:#?}", tokens);
    match tokens.remove(0) {
        TokenData::Word(name) => {
            if name == "return" {
                let expr = parse_expression(tokens)?;
                return Ok(Statement::Return { expr });
            } else if tokens.first() == Some(&TokenData::Symbols("=".to_string())) {
                tokens.remove(0);
                trace!("Parsing assignment statement");
                let value = parse_expression(tokens)?;
                return Ok(Statement::Assign {
                    variable: name,
                    value,
                });
            } else {
                let expr = parse_expression(tokens)?;
                return Ok(Statement::Expression { expr });
            }
        }
        token => return Err(anyhow::anyhow!("expected statement instead of {:?}", token)),
    }
}

fn parse_function_args(tokens: &mut Vec<TokenData>) -> anyhow::Result<Vec<Parameter>> {
    let mut parameters = vec![];
    loop {
        match tokens.first().unwrap() {
            TokenData::Symbols(s) => {
                if s == "," {
                    tokens.remove(0);
                    continue;
                } else {
                    return Err(anyhow::anyhow!("expected ',', parameter name, or ')'"));
                }
            }
            TokenData::Parenthesis {
                opening: ')',
                is_opening: false,
            } => {
                tokens.remove(0);
                break;
            }
            TokenData::Word(name) => {
                let name = name.clone();
                anyhow::ensure!(
                    tokens.remove(0) == TokenData::Symbols(":".to_string()),
                    "expected ':' after parameter name"
                );
                let ty = match tokens.remove(0) {
                    TokenData::Word(ty) => ty,
                    _ => return Err(anyhow::anyhow!("expected parameter type")),
                };

                parameters.push(Parameter { name, ty });
            }
            _ => return Err(anyhow::anyhow!("expected parameter name")),
        }
    }

    Ok(parameters)
}

fn parse_function_declaration(tokens: &mut Vec<TokenData>) -> anyhow::Result<Function> {
    anyhow::ensure!(
        tokens.remove(0) == TokenData::Word("fun".to_string()),
        "expected 'fun' keyword"
    );

    let name = match tokens.remove(0) {
        TokenData::Word(name) => name,
        _ => return Err(anyhow::anyhow!("expected function name")),
    };

    anyhow::ensure!(
        tokens.remove(0)
            == TokenData::Parenthesis {
                opening: '(',
                is_opening: true
            },
        "expected opening parenthesis"
    );

    let parameters = parse_function_args(tokens)?;

    anyhow::ensure!(
        tokens.remove(0)
            == TokenData::Parenthesis {
                opening: '{',
                is_opening: true
            },
        "expected opening brace"
    );

    anyhow::ensure!(
        tokens.remove(0) == TokenData::Newline,
        "expected newline after opening brace"
    );

    let mut body = vec![];
    loop {
        trace!(
            "Parsing entry in a function body: {:#?}",
            tokens.first().unwrap()
        );
        match tokens.first().unwrap() {
            TokenData::Parenthesis {
                opening: '}',
                is_opening: false,
            } => {
                tokens.remove(0);
                break;
            }
            TokenData::Newline => {
                tokens.remove(0);
            }
            _ => {
                let statement = parse_statement(tokens)?;
                body.push(statement);
            }
        }
    }

    Ok(Function {
        name,
        parameters,
        body,
    })
}

#[tracing::instrument(level = "trace", skip(tokens))]
pub fn parse(mut tokens: TokenList) -> anyhow::Result<Ast> {
    // Discard all comments.
    tokens
        .tokens
        .retain(|token| !matches!(token, TokenData::Comment(_)));

    let mut ast = Ast {
        top_level_functions: vec![],
    };

    while let Some(token) = tokens.tokens.first() {
        match token {
            TokenData::Word(name) if name == "fun" => {
                let function = parse_function_declaration(&mut tokens.tokens)?;
                ast.top_level_functions.push(function);
            }
            TokenData::Newline => {
                tokens.tokens.remove(0);
            }
            _ => return Err(anyhow::anyhow!("unexpected token {:?}", token)),
        }
    }

    Ok(ast)
}

fn parse_expression(tokens: &mut Vec<TokenData>) -> anyhow::Result<Expression> {
    // Parse literal, variable, or function call
    let first_token = tokens.remove(0);
    match (first_token, tokens.first()) {
        (
            TokenData::Parenthesis {
                opening: '(',
                is_opening: true,
            },
            _,
        ) => {
            trace!("Parsing expression");
            tokens.remove(0);
            let expr = parse_expression(tokens)?;
            anyhow::ensure!(
                tokens.remove(0)
                    == TokenData::Parenthesis {
                        opening: ')',
                        is_opening: false
                    },
                "Expected closing parenthesis"
            );
            Ok(expr)
        }
        (
            TokenData::Word(name),
            Some(TokenData::Parenthesis {
                opening: '(',
                is_opening: true,
            }),
        ) => {
            tokens.remove(0);
            let mut args = vec![];
            loop {
                match tokens.get(0) {
                    Some(TokenData::Parenthesis {
                        opening: ')',
                        is_opening: false,
                    }) => {
                        tokens.remove(0);
                        break;
                    }
                    _ => {
                        let expr = parse_expression(tokens)?;
                        args.push(expr);
                        // TODO: comma and more args
                    }
                }
            }
            Ok(Expression::Call(name, args))
        }
        (TokenData::Word(name), _) => Ok(Expression::Variable(name)),
        (TokenData::String(s), _) => Ok(Expression::Literal(Literal::String(s))),
        (TokenData::Number(value), _) => Ok(Expression::Literal(Literal::Int(value))),
        (token, _) => Err(anyhow::anyhow!("unexpected token {:?}", token)),
    }
}
