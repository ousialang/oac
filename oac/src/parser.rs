use super::lexer::TokenList;

#[derive(Debug)]
pub struct SyntaxTree {
    nodes: Vec<SyntaxNode>,
}

#[derive(Debug)]
pub enum SyntaxNode {
    Constant {
        identifier: String,
        typ: String,
        value: String,
    },
    Variable {
        identifier: String,
        typ: String,
        value: String,
    },
    SetValue {
        identifier: String,
        value: String,
    },
    Print {
        identifire: String,
    },
}

pub fn parse(tokens: TokenList) -> AST {}
