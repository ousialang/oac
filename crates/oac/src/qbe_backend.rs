use qbe::*;

use crate::parser::Ast;

pub fn compile(ast: Ast) -> qbe::Module<'static> {
    let mut module = qbe::Module::default();

    module.add_data(qbe::DataDef::new(
        Linkage::private(),
        "fmt".to_string(),
        None,
        vec![
            (qbe::Type::Byte, DataItem::Str("%s\\n\\n".to_string())),
            (qbe::Type::Byte, DataItem::Const(0)),
        ],
    ));

    module.add_data(qbe::DataDef::new(
        Linkage::private(),
        "string_contents".to_string(),
        None,
        vec![
            (qbe::Type::Byte, DataItem::Str("foobar".to_string())),
            (qbe::Type::Byte, DataItem::Const(0)),
        ],
    ));

    let mut func = qbe::Function::new(
        qbe::Linkage::public(),
        "main".to_string(),
        vec![],
        Some(qbe::Type::Word),
    );
    func.add_block("start".to_string());
    func.add_instr(qbe::Instr::Call(
        "printf".into(),
        vec![
            (qbe::Type::Long, qbe::Value::Global("fmt".into())),
            (
                qbe::Type::Long,
                qbe::Value::Global("string_contents".into()),
            ),
        ],
    ));
    func.add_instr(qbe::Instr::Ret(Some(qbe::Value::Const(0))));

    module.add_function(func);
    module
}
