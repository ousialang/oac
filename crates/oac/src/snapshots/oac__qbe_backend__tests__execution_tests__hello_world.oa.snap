---
source: crates/oac/src/qbe_backend.rs
expression: output
---

